File: libraries/maui-iap/src/OpenIap.Maui/KitApi.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;
using OpenIap;

namespace OpenIap.Maui;

/// <summary>
/// Options for the OpenIAP kit HTTP API. Mirrors the JavaScript
/// <c>kitApi({ apiKey, baseUrl })</c> helper.
/// </summary>
public sealed record KitApiOptions
{
    public required string ApiKey { get; init; }
    public string? BaseUrl { get; init; }
    public HttpClient? HttpClient { get; init; }
}

internal static class TolerantJson
{
    public static T? Deserialize<T>(JsonElement element, JsonSerializerOptions options)
        where T : class
    {
        try
        {
            return element.Deserialize<T>(options);
        }
        catch (JsonException)
        {
            return null;
        }
        catch (InvalidOperationException)
        {
            return null;
        }
        catch (NotSupportedException)
        {
            return null;
        }
    }
}

internal sealed class TolerantJsonObjectConverter<T> : JsonConverter<T?>
    where T : class
{
    public override T? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        using var document = JsonDocument.ParseValue(ref reader);
        return TolerantJson.Deserialize<T>(document.RootElement, options);
    }

    public override void Write(Utf8JsonWriter writer, T? value, JsonSerializerOptions options)
        => JsonSerializer.Serialize(writer, value, options);
}

internal sealed class TolerantJsonListConverter<T> : JsonConverter<IReadOnlyList<T>>
    where T : class
{
    public override IReadOnlyList<T> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        using var document = JsonDocument.ParseValue(ref reader);
        if (document.RootElement.ValueKind != JsonValueKind.Array)
        {
            throw new JsonException($"Expected an array of {typeof(T).Name} values.");
        }

        var values = new List<T>();
        foreach (var element in document.RootElement.EnumerateArray())
        {
            var value = TolerantJson.Deserialize<T>(element, options);
            if (value is not null) values.Add(value);
        }
        return values;
    }

    public override void Write(Utf8JsonWriter writer, IReadOnlyList<T> value, JsonSerializerOptions options)
        => JsonSerializer.Serialize(writer, value, options);
}

public sealed record KitSubscription
{
    [JsonPropertyName("id")]
    public required string Id { get; init; }
    [JsonPropertyName("productId")]
    public required string ProductId { get; init; }
    [JsonPropertyName("platform")]
    public required IapPlatform Platform { get; init; }
    [JsonPropertyName("state")]
    public required string State { get; init; }
    [JsonPropertyName("expiresAt")]
    public double? ExpiresAt { get; init; }
    [JsonPropertyName("renewsAt")]
    public double? RenewsAt { get; init; }
    [JsonPropertyName("willRenew")]
    public bool? WillRenew { get; init; }
    [JsonPropertyName("cancellationReason")]
    public string? CancellationReason { get; init; }
    [JsonPropertyName("currency")]
    public string? Currency { get; init; }
    [JsonPropertyName("priceAmountMicros")]
    public double? PriceAmountMicros { get; init; }
    [JsonPropertyName("startedAt")]
    public required double StartedAt { get; init; }
    [JsonPropertyName("updatedAt")]
    public required double UpdatedAt { get; init; }
    [JsonPropertyName("purchaseToken")]
    public required string PurchaseToken { get; init; }
    [JsonPropertyName("userId")]
    public string? UserId { get; init; }
}

public sealed record EntitlementsResponse
{
    [JsonPropertyName("userId")]
    public required string UserId { get; init; }
    [JsonPropertyName("productIds")]
    public required IReadOnlyList<string> ProductIds { get; init; }
    [JsonPropertyName("subscriptions")]
    [JsonConverter(typeof(TolerantJsonListConverter<KitSubscription>))]
    public required IReadOnlyList<KitSubscription> Subscriptions { get; init; }
}

public sealed record StatusResponse
{
    [JsonPropertyName("active")]
    public required bool Active { get; init; }
    [JsonPropertyName("subscription")]
    [JsonConverter(typeof(TolerantJsonObjectConverter<KitSubscription>))]
    public KitSubscription? Subscription { get; init; }
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum KitProductPlatform
{
    IOS,
    Android,
}

public sealed record KitProductClientPayload
{
    [JsonPropertyName("format")]
    public required string Format { get; init; }
    [JsonPropertyName("body")]
    public required string Body { get; init; }
    [JsonPropertyName("version")]
    public required double Version { get; init; }
    [JsonPropertyName("updatedAt")]
    public required double UpdatedAt { get; init; }
}

public sealed record KitProductOffer
{
    [JsonPropertyName("id")]
    public required string Id { get; init; }
    [JsonPropertyName("kind")]
    public required string Kind { get; init; }
    [JsonPropertyName("duration")]
    public string? Duration { get; init; }
    [JsonPropertyName("numberOfPeriods")]
    public double? NumberOfPeriods { get; init; }
    [JsonPropertyName("priceAmountMicros")]
    public double? PriceAmountMicros { get; init; }
    [JsonPropertyName("currency")]
    public string? Currency { get; init; }
}

public sealed record KitProduct
{
    [JsonPropertyName("productId")]
    public required string ProductId { get; init; }
    [JsonPropertyName("platform")]
    public required KitProductPlatform Platform { get; init; }
    [JsonPropertyName("type")]
    public required string Type { get; init; }
    [JsonPropertyName("title")]
    public required string Title { get; init; }
    [JsonPropertyName("description")]
    public string? Description { get; init; }
    [JsonPropertyName("priceAmountMicros")]
    public double? PriceAmountMicros { get; init; }
    [JsonPropertyName("currency")]
    public string? Currency { get; init; }
    [JsonPropertyName("state")]
    public required string State { get; init; }
    [JsonPropertyName("storeRef")]
    public string? StoreRef { get; init; }
    [JsonPropertyName("subscriptionGroupId")]
    public string? SubscriptionGroupId { get; init; }
    [JsonPropertyName("subscriptionGroupName")]
    public string? SubscriptionGroupName { get; init; }
    [JsonPropertyName("billingPeriod")]
    public string? BillingPeriod { get; init; }
    [JsonPropertyName("offers")]
    [JsonConverter(typeof(TolerantJsonListConverter<KitProductOffer>))]
    public IReadOnlyList<KitProductOffer>? Offers { get; init; }
    [JsonPropertyName("updatedAt")]
    public required double UpdatedAt { get; init; }
    [JsonPropertyName("clientPayload")]
    public KitProductClientPayload? ClientPayload { get; init; }
}

public sealed record KitProductsOptions
{
    public KitProductPlatform? Platform { get; init; }
    public bool? IncludeClientPayload { get; init; }
    public int? Limit { get; init; }
    public string? Cursor { get; init; }
}

public sealed record KitProductsResponse
{
    [JsonPropertyName("products")]
    [JsonConverter(typeof(TolerantJsonListConverter<KitProduct>))]
    public required IReadOnlyList<KitProduct> Products { get; init; }
    [JsonPropertyName("hasMore")]
    public bool? HasMore { get; init; }
    [JsonPropertyName("nextCursor")]
    public string? NextCursor { get; init; }
}

public sealed record KitClientPayloadResponse
{
    [JsonPropertyName("clientPayload")]
    public required KitProductClientPayload ClientPayload { get; init; }
}

public sealed record BindUserResponse
{
    [JsonPropertyName("ok")]
    public required bool Ok { get; init; }
    [JsonPropertyName("bound")]
    public required bool Bound { get; init; }
}

public sealed class KitApiError : Exception
{
    public KitApiError(int status, object? body, string message)
        : base(message)
    {
        Status = status;
        Body = body;
    }

    public int Status { get; }
    public object? Body { get; }
}

public sealed class KitApiClient
{
    private const string DefaultBaseUrl = "https://kit.openiap.dev";
    private readonly HttpClient _http;

    internal KitApiClient(KitApiOptions options)
    {
        if (string.IsNullOrWhiteSpace(options.ApiKey))
        {
            throw new ArgumentException("kitApi requires a non-empty apiKey.", nameof(options));
        }

        ApiKey = options.ApiKey;
        BaseUrl = TrimTrailingSlash(options.BaseUrl ?? DefaultBaseUrl);
        _http = options.HttpClient ?? new HttpClient();
    }

    public string ApiKey { get; }
    public string BaseUrl { get; }

    /// <summary>
    /// GET /v1/subscriptions/status/{apiKey}?userId=...
    /// </summary>
    public Task<StatusResponse> StatusAsync(string userId, CancellationToken cancellationToken = default)
        => CallAsync<StatusResponse>(
            $"/v1/subscriptions/status/{Uri.EscapeDataString(ApiKey)}?userId={Uri.EscapeDataString(userId)}",
            null,
            cancellationToken);

    /// <summary>
    /// GET /v1/subscriptions/entitlements/{apiKey}?userId=...
    /// </summary>
    public Task<EntitlementsResponse> EntitlementsAsync(string userId, CancellationToken cancellationToken = default)
        => CallAsync<EntitlementsResponse>(
            $"/v1/subscriptions/entitlements/{Uri.EscapeDataString(ApiKey)}?userId={Uri.EscapeDataString(userId)}",
            null,
            cancellationToken);

    /// <summary>
    /// GET /v1/products/{apiKey}. Client payload bodies are opt-in.
    /// </summary>
    public Task<KitProductsResponse> ProductsAsync(
        KitProductsOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        if (options?.IncludeClientPayload == true && options.Platform is null)
        {
            throw new ArgumentException(
                "ProductsAsync requires Platform when IncludeClientPayload is true.",
                nameof(options));
        }
        var query = new List<string>();
        if (options?.Platform is { } platform)
        {
            query.Add($"platform={Uri.EscapeDataString(platform.ToString())}");
        }
        if (options?.IncludeClientPayload is { } includeClientPayload)
        {
            query.Add($"includeClientPayload={includeClientPayload.ToString().ToLowerInvariant()}");
        }
        if (options?.Limit is { } limit)
        {
            query.Add($"limit={limit}");
        }
        if (options?.Cursor is { } cursor)
        {
            query.Add($"cursor={Uri.EscapeDataString(cursor)}");
        }
        var suffix = query.Count == 0 ? string.Empty : $"?{string.Join("&", query)}";
        return CallAsync<KitProductsResponse>(
            $"/v1/products/{Uri.EscapeDataString(ApiKey)}{suffix}",
            null,
            cancellationToken);
    }

    /// <summary>
    /// GET one public client payload by (platform, productId).
    /// </summary>
    public Task<KitClientPayloadResponse> ClientPayloadAsync(
        string productId,
        KitProductPlatform platform,
        CancellationToken cancellationToken = default)
        => CallAsync<KitClientPayloadResponse>(
            $"/v1/products/{Uri.EscapeDataString(ApiKey)}/{Uri.EscapeDataString(productId)}/client-payload?platform={Uri.EscapeDataString(platform.ToString())}",
            null,
            cancellationToken);

    /// <summary>
    /// POST /v1/subscriptions/bind-user/{apiKey}
    /// </summary>
    public Task<BindUserResponse> BindUserAsync(
        string purchaseToken,
        string userId,
        CancellationToken cancellationToken = default)
    {
        var body = JsonSerializer.Serialize(new { purchaseToken, userId }, JsonOptions.Default);
        return CallAsync<BindUserResponse>(
            $"/v1/subscriptions/bind-user/{Uri.EscapeDataString(ApiKey)}",
            new StringContent(body, Encoding.UTF8, "application/json"),
            cancellationToken);
    }

    private async Task<T> CallAsync<T>(
        string path,
        HttpContent? content,
        CancellationToken cancellationToken)
    {
        var normalizedPath = path.StartsWith("/", StringComparison.Ordinal) ? path : $"/{path}";
        using var request = new HttpRequestMessage(
            content is null ? HttpMethod.Get : HttpMethod.Post,
            $"{BaseUrl}{normalizedPath}")
        {
            Content = content,
        };
        request.Headers.Accept.ParseAdd("application/json");

        using var response = await _http.SendAsync(request, cancellationToken).ConfigureAwait(false);
        var text = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
        var parsed = ParseBody(text, out var parseError);

        if (!response.IsSuccessStatusCode)
        {
            throw new KitApiError(
                (int)response.StatusCode,
                parsed ?? text,
                $"kit {path} returned {(int)response.StatusCode}");
        }

        if (parseError is not null)
        {
            throw new KitApiError(
                (int)response.StatusCode,
                text,
                $"kit {path} returned a non-JSON {(int)response.StatusCode} body ({parseError.Message})");
        }

        if (parsed is null)
        {
            throw new KitApiError(
                (int)response.StatusCode,
                null,
                $"kit {path} returned an empty {(int)response.StatusCode} body");
        }

        return parsed.Deserialize<T>(JsonOptions.Default)
            ?? throw new KitApiError(
                (int)response.StatusCode,
                parsed,
                $"kit {path} returned an unexpected body");
    }

    private static JsonNode? ParseBody(string text, out JsonException? parseError)
    {
        parseError = null;
        if (string.IsNullOrEmpty(text)) return null;
        try
        {
            return JsonNode.Parse(text);
        }
        catch (JsonException ex)
        {
            parseError = ex;
            return null;
        }
    }

    internal static string TrimTrailingSlash(string url)
        => url.EndsWith("/", StringComparison.Ordinal) ? url[..^1] : url;
}

File: libraries/react-native-iap/src/kit-api.ts
// Tiny fetch wrapper around kit's `/v1` HTTP surface for use by the JS
// SDK consumers (react-native-iap + expo-iap). Mirrors the shape of
// `packages/mcp-server/src/kit-client.ts` so the same operations are
// reachable from both LLM tools and end-user apps without each
// duplicating the URL layout.

export type KitApiOptions = {
  apiKey: string;
  baseUrl?: string;
  // Optional fetch override for runtimes without a global (older RN
  // builds) or for injection in tests.
  fetchImpl?: (input: string, init?: RequestInit) => Promise<Response>;
  /** Optional AsyncStorage-compatible persistent cache for direct client
   * payload reads. Cache failures never change a successful API result. */
  clientPayloadCache?: KitClientPayloadCache;
};

export type KitClientPayloadCache = {
  getItem: (key: string) => Promise<string | null> | string | null;
  setItem: (key: string, value: string) => Promise<void> | void;
  removeItem?: (key: string) => Promise<void> | void;
};

export type KitClientPayloadOptions = {
  /** Revalidate a cached payload with its scoped ETag. Without this flag,
   * a valid persistent entry is returned without a network request. */
  refresh?: boolean;
};

export type KitSubscription = {
  id: string;
  productId: string;
  platform: "IOS" | "Android";
  state: string;
  expiresAt?: number;
  renewsAt?: number;
  willRenew?: boolean;
  cancellationReason?: string;
  currency?: string;
  priceAmountMicros?: number;
  startedAt: number;
  updatedAt: number;
  purchaseToken: string;
  userId?: string;
};

export type EntitlementsResponse = {
  userId: string;
  productIds: string[];
  subscriptions: KitSubscription[];
};

export type StatusResponse = {
  active: boolean;
  subscription: KitSubscription | null;
};

export type KitProductPlatform = "IOS" | "Android";

export type KitProductClientPayload = {
  format: "toml" | "json" | "text";
  body: string;
  version: number;
  updatedAt: number;
};

export type KitProductOffer = {
  id: string;
  kind:
    | "FreeTrial"
    | "IntroPayUpFront"
    | "IntroPayAsYouGo"
    | "PromotionalOffer"
    | "BasePlan";
  duration?: string;
  numberOfPeriods?: number;
  priceAmountMicros?: number;
  currency?: string;
};

export type KitProduct = {
  productId: string;
  platform: KitProductPlatform;
  type: "Subscription" | "NonConsumable" | "Consumable";
  title: string;
  description?: string;
  priceAmountMicros?: number;
  currency?: string;
  state: "Draft" | "Ready" | "Active" | "Removed";
  storeRef?: string;
  subscriptionGroupId?: string;
  subscriptionGroupName?: string;
  billingPeriod?: "P1W" | "P1M" | "P2M" | "P3M" | "P6M" | "P1Y";
  offers?: KitProductOffer[];
  updatedAt: number;
  clientPayload?: KitProductClientPayload;
};

export type KitProductsOptions = {
  platform?: KitProductPlatform;
  /** Include public client payload bodies in a bounded platform page. */
  includeClientPayload?: boolean;
  /** Page size for every catalog read (default 25, maximum 50). */
  limit?: number;
  /** Opaque cursor returned as `nextCursor` by the previous page. */
  cursor?: string;
};

export type KitProductsResponse = {
  products: KitProduct[];
  hasMore: boolean;
  /** Present when a catalog page has another page. */
  nextCursor?: string;
};

export type KitClientPayloadResponse = {
  clientPayload: KitProductClientPayload;
};

type CachedClientPayload = KitClientPayloadResponse & {
  etag?: string;
};

type InternalRequestInit = Omit<RequestInit, "headers"> & {
  headers?: Record<string, string>;
};

const DEFAULT_BASE_URL = "https://kit.openiap.dev";

// Merge the request's internal headers with kit defaults (`accept`,
// optionally `content-type`). When `Headers` is missing — older React
// Native builds where the operator wires up `fetchImpl` without a
// `Headers` polyfill — the internal request sites use plain records,
// so a small case-insensitive merge is sufficient.
function mergeHeaders(
  callerHeaders: Record<string, string> | undefined,
  hasBody: boolean,
): HeadersInit {
  if (typeof Headers === "function") {
    const merged = new Headers(callerHeaders);
    if (!merged.has("accept")) merged.set("accept", "application/json");
    if (hasBody && !merged.has("content-type")) {
      merged.set("content-type", "application/json");
    }
    return merged;
  }
  // Plain-object fallback path. Build a case-insensitive name map and
  // re-emit it as a record `fetchImpl` accepts.
  const lower = new Map<string, { name: string; value: string }>();
  const setIfAbsent = (name: string, value: string) => {
    const key = name.toLowerCase();
    if (!lower.has(key)) lower.set(key, { name, value });
  };
  if (callerHeaders) {
    for (const [name, value] of Object.entries(callerHeaders)) {
      lower.set(name.toLowerCase(), { name, value });
    }
  }
  setIfAbsent("accept", "application/json");
  if (hasBody) setIfAbsent("content-type", "application/json");
  const out: Record<string, string> = {};
  for (const { name, value } of lower.values()) out[name] = value;
  return out;
}

export class KitApiError extends Error {
  constructor(
    readonly status: number,
    readonly body: unknown,
    message: string,
  ) {
    super(message);
    this.name = "KitApiError";
  }
}

export function kitApi(options: KitApiOptions) {
  const baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/$/, "");
  const fetchImpl: (input: string, init?: RequestInit) => Promise<Response> =
    (() => {
      if (options.fetchImpl) return options.fetchImpl;
      if (typeof fetch === "function") {
        return (input: string, init?: RequestInit) => fetch(input, init);
      }
      throw new Error(
        "kitApi requires a fetch implementation. Pass `fetchImpl` for runtimes without a global fetch.",
      );
    })();

  async function request(
    path: string,
    init?: InternalRequestInit,
  ): Promise<Response> {
    // Normalize headers without depending on a global `Headers`
    // constructor: older React Native runtimes ship `fetch` (or a
    // polyfill via `fetchImpl`) without exposing `Headers` globally.
    // The prior implementation crashed before the first request on
    // those runtimes. We use `new Headers()` when available and
    // otherwise fall back to a small case-insensitive merge into a
    // plain record. Either way, kit defaults only apply when the
    // internal request hasn't set the same name.
    const headers = mergeHeaders(init?.headers, init?.body != null);
    // Prepend a leading slash if `path` is missing one. Today's
    // call sites all hard-code the leading "/", but normalizing here
    // makes the helper safe for future additions and matches the
    // already-stripped `baseUrl` (PR #124
    // (https://github.com/hyodotdev/openiap/pull/124) review).
    const normalizedPath = path.startsWith("/") ? path : `/${path}`;
    return fetchImpl(`${baseUrl}${normalizedPath}`, {
      ...init,
      headers,
    });
  }

  async function parseResponse<T>(
    response: Response,
    path: string,
  ): Promise<T> {
    const text = await response.text();
    // Empty body normalizes to null so callers expecting JSON
    // (status / entitlements / list*) don't get a truthy ""
    // and crash on property access.
    let parsed: unknown = null;
    let parseError: unknown = null;
    if (text) {
      try {
        parsed = JSON.parse(text);
      } catch (error) {
        // Non-JSON body (a misconfigured proxy returning HTML, a
        // CDN-injected error page, etc.) on a 2xx response would
        // otherwise reach the caller as `parsed = text` and crash
        // on property access via `parsed as T`. Throw a structured
        // KitApiError instead so callers see a typed failure.
        parseError = error;
      }
    }
    if (!response.ok) {
      // Surface the raw body (text or parsed) on the error path so
      // operators can read the upstream error message verbatim.
      throw new KitApiError(
        response.status,
        parsed ?? text,
        `kit ${path} returned ${response.status}`,
      );
    }
    if (parseError) {
      throw new KitApiError(
        response.status,
        text,
        `kit ${path} returned a non-JSON ${response.status} body (${
          parseError instanceof Error ? parseError.message : String(parseError)
        })`,
      );
    }
    return parsed as T;
  }

  async function call<T>(path: string, init?: InternalRequestInit): Promise<T> {
    return parseResponse<T>(await request(path, init), path);
  }

  function clientPayloadCacheKey(
    productId: string,
    platform: KitProductPlatform,
  ): string {
    return [
      "iapkit-client-payload-v1",
      baseUrl,
      options.apiKey,
      platform,
      productId,
    ]
      .map(encodeURIComponent)
      .join(":");
  }

  async function readCachedClientPayload(
    cacheKey: string,
  ): Promise<CachedClientPayload | null> {
    if (!options.clientPayloadCache) return null;
    try {
      const raw = await options.clientPayloadCache.getItem(cacheKey);
      if (!raw) return null;
      const candidate = JSON.parse(raw) as Partial<CachedClientPayload>;
      const payload = candidate.clientPayload;
      // Only the invariants the cache itself depends on. `format` is opaque
      // here: rejecting a format IAPKit added later would evict the entry,
      // stop the ETag revalidation below from ever being sent, and leave
      // offline reads with nothing — for a value the live path passes through
      // to the caller unchanged anyway.
      if (
        !payload ||
        typeof payload.format !== "string" ||
        typeof payload.body !== "string" ||
        !Number.isSafeInteger(payload.version) ||
        payload.version < 1 ||
        typeof payload.updatedAt !== "number" ||
        (candidate.etag !== undefined && typeof candidate.etag !== "string")
      ) {
        await options.clientPayloadCache.removeItem?.(cacheKey);
        return null;
      }
      return candidate as CachedClientPayload;
    } catch {
      return null;
    }
  }

  async function writeCachedClientPayload(
    cacheKey: string,
    value: CachedClientPayload,
  ): Promise<void> {
    try {
      await options.clientPayloadCache?.setItem(
        cacheKey,
        JSON.stringify(value),
      );
    } catch {
      // Persistence is an optimization. A device storage failure must not
      // turn a successful IAPKit read into an application error.
    }
  }

  async function removeCachedClientPayload(cacheKey: string): Promise<void> {
    try {
      await options.clientPayloadCache?.removeItem?.(cacheKey);
    } catch {
      // See writeCachedClientPayload: cache maintenance is best effort.
    }
  }

  return {
    apiKey: options.apiKey,
    baseUrl,

    /** GET /v1/subscriptions/status — the `active` boolean is the
     * fastest gate for "is this user paying?". */
    status: (userId: string) =>
      call<StatusResponse>(
        `/v1/subscriptions/status/${encodeURIComponent(options.apiKey)}?userId=${encodeURIComponent(userId)}`,
      ),

    /** GET /v1/subscriptions/entitlements — every productId the user
     * is entitled to. Use this when feature gating depends on which
     * specific tier the user owns. */
    entitlements: (userId: string) =>
      call<EntitlementsResponse>(
        `/v1/subscriptions/entitlements/${encodeURIComponent(options.apiKey)}?userId=${encodeURIComponent(userId)}`,
      ),

    /** GET /v1/products — read the store-synced catalog. Client payloads
     * are excluded unless explicitly requested because they may add up to
     * 16 KiB per product. */
    products: (productOptions: KitProductsOptions = {}) => {
      if (
        productOptions.includeClientPayload === true &&
        !productOptions.platform
      ) {
        throw new Error(
          "kitApi.products requires platform when includeClientPayload is true",
        );
      }
      const query = new URLSearchParams();
      if (productOptions.platform) {
        query.set("platform", productOptions.platform);
      }
      if (productOptions.includeClientPayload !== undefined) {
        query.set(
          "includeClientPayload",
          String(productOptions.includeClientPayload),
        );
      }
      if (productOptions.limit !== undefined) {
        query.set("limit", String(productOptions.limit));
      }
      if (productOptions.cursor !== undefined) {
        query.set("cursor", productOptions.cursor);
      }
      const encodedQuery = query.toString();
      const suffix = encodedQuery ? `?${encodedQuery}` : "";
      return call<KitProductsResponse>(
        `/v1/products/${encodeURIComponent(options.apiKey)}${suffix}`,
      );
    },

    /** GET one public client payload by its store-specific natural key.
     * Payloads are app-facing data; never store secrets in them. */
    clientPayload: async (
      productId: string,
      platform: KitProductPlatform,
      payloadOptions: KitClientPayloadOptions = {},
    ) => {
      const path = `/v1/products/${encodeURIComponent(options.apiKey)}/${encodeURIComponent(productId)}/client-payload?platform=${encodeURIComponent(platform)}`;
      const cacheKey = clientPayloadCacheKey(productId, platform);
      const cached = await readCachedClientPayload(cacheKey);
      if (cached && payloadOptions.refresh !== true) {
        return { clientPayload: cached.clientPayload };
      }
      const response = await request(path, {
        ...(cached?.etag ? { headers: { "If-None-Match": cached.etag } } : {}),
      });
      if (response.status === 304 && cached) {
        return { clientPayload: cached.clientPayload };
      }
      try {
        const result = await parseResponse<KitClientPayloadResponse>(
          response,
          path,
        );
        await writeCachedClientPayload(cacheKey, {
          ...result,
          ...(response.headers.get("etag")
            ? { etag: response.headers.get("etag")! }
            : {}),
        });
        return result;
      } catch (error) {
        if (error instanceof KitApiError && error.status === 404) {
          await removeCachedClientPayload(cacheKey);
        }
        throw error;
      }
    },

    /** POST /v1/subscriptions/bind-user — call after a successful
     * verifyReceipt so kit knows which userId owns the verified
     * `purchaseToken`. Idempotent. */
    bindUser: (purchaseToken: string, userId: string) =>
      call<{ ok: boolean; bound: boolean }>(
        `/v1/subscriptions/bind-user/${encodeURIComponent(options.apiKey)}`,
        {
          method: "POST",
          body: JSON.stringify({ purchaseToken, userId }),
        },
      ),
  };
}
